### ISSUE TITLE
Refresh materialized views concurrently

### ISSUE THREAD

**Issue by eljhkrr** [medic/cht-app-monitoring-data-ingest#52]
Non-concurrent refresh is causing dashboard downtime lasting several hours:
![Screenshot from 2023-02-21 18-33-01](https://user-images.githubusercontent.com/4529018/220512507-71f5bcdb-a7f5-46e0-925d-75da7a8d5d5d.png)

**Comment by jdndiaye** [comment]
Hi @Yuvraj, @eljhkrr 
I would like start to work on this issue. Can you give me some guidelines so that I can get my bearings?
Is it a problem with query performance? Or is it something you can fix on superset?

**Comment by yrimal** [comment]
Hi @eljhkrr,
Can you share why views were changed not to refresh concurrently in the beginning?

**Comment by freddieptf** [comment]
@jdndiaye it's from a discussion in this [PR](https://github.com/medic/app-services-team/pull/388), concurrent refreshes won't block selects but will take longer. I don't think the pattern we're currently using is scalable as the data and time it will take to refresh will only grow. @yrimal had mentioned he uses tables instead of materialized views for some of the impact dashboards, i think that's worth exploring? Maybe we could use the last refresh date to get only new data? What do you think @jdndiaye?

**Comment by yrimal** [comment]
Yes, that's the one way of doing it. Do you know how many data sources are being used by superset dashboard? 

I planned impact dashboard from the beginning and it was easier. However, in app monitoring's case, since we've already built many charts based on these views, Changing source for all of these dashboards might be too hectic. If you can try that and check, please do so. Otherwise here's another alternative I've thought that'll require no change on superset side. This is the mechanism to staged refreshing. For example, let's consider a materialized view `app_monitoring_couchpg_backlog`. 

1. Refresh materialized views as `_staging`. For example, create a materialized view called `app_monitoring_couchpg_backlog_staging` with the same source as `app_monitoring_couchpg_backlog`. 
2. Do this for all materialized views. 
3. When refreshing is completed, swap the view names. i.e. Rename `app_monitoring_couchpg_backlog_staging` view to `app_monitoring_couchpg_backlog`. This way, superset always points to `app_monitoring_couchpg_backlog` without any blockade. 

Feel free to pick one of the option among these. Rename materialized views can go as one script after all refreshing is completed.

**Comment by jdndiaye** [comment]
Thank you @yrimal, I'm going with creating news materialized views after all refreshing is completed I'll rename them.

**Comment by jdndiaye** [comment]
Hi @eljhkrr @yrimal All materialized views for app_monitoring_* are refreshed. The dashboards seem to work well. Please check and get back to me.

**Comment by jdndiaye** [comment]
@freddieptf The solution we came up with doesn't seem to work ([comment](https://medic.slack.com/archives/C03CYKC7NUD/p1688642550462909?thread_ts=1688382162.430989&cid=C03CYKC7NUD)). So should we try the other alternative which is to use tables instead of materialized views? But as @yrimal pointed out, changing source for all of these dashboards might be too hectic.

**Comment by freddieptf** [comment]
@jdndiaye the solution will work, as stated [above](https://github.com/medic/cht-app-monitoring-data-ingest/issues/52#issuecomment-1594231869)

- update refresh function to only refresh staging materialized views
- create staging materialized views
- When refresh is complete, swap the view names. eg rename `app_monitoring_matview_staging` view to `app_monitoring_matview` and `app_monitoring_matview` to `app_monitoring_matview_staging` so the old view is refreshed the next time the refresh function is called. You might have to use some extra steps while renaming to avoid name collisions. 

If you want to save on space after the refresh is done i think you can call `REFRESH MATERIALIZED VIEW matview_name WITH NO DATA;` to [free space](https://www.postgresql.org/docs/current/sql-refreshmaterializedview.html) associated with the old materialized view

**Comment by freddieptf** [comment]
These are some stats on refresh times from our matview logs
|view_name|elapsed_time|
|---------|------------|
|app_monitoring_chw_last_sync|01:44:38.922987|
|app_monitoring_tasks_by_state|01:38:26.077638|
|app_monitoring_tasks_by_user|01:37:49.601313|
|app_monitoring_invalid_doc_errors|01:28:54.279848|
|app_monitoring_invalid_docs|01:16:43.194798|
|app_monitoring_purge_logs|00:34:44.004702|
|app_monitoring_form_completions|00:27:03.804456|
|app_monitoring_replication_by_status|00:26:29.946792|
|app_monitoring_replication_failure_reasons|00:26:25.93651|
|app_monitoring_sms_outgoing_states|00:12:34.368485|
|app_monitoring_feedback_docs|00:07:21.951576|
|app_monitoring_users_with_insufficient_chrome|00:03:23.654716|
|app_monitoring_users_with_insufficient_disk|00:03:13.008847|
|app_monitoring_upgrade_events|00:01:26.946953|
|app_monitoring_initial_replications|00:01:09.528938|
|app_monitoring_klipfolio_datasource_refresh|00:00:21.004922|
|app_monitoring_users_replication_limit|00:00:00.797816|
|app_monitoring_sms_error_and_users_docs_replication|00:00:00.275443|
|app_monitoring_sentinel_backlog|00:00:00.272715|
|app_monitoring_outbound_backlog|00:00:00.10395|
|app_monitoring_chw_engagement|00:00:00.092342|

I'm going to run some explains on the query backing `app_monitoring_chw_last_sync` and see if i can optimize it
